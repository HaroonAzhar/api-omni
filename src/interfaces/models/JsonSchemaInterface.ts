export interface JsonSchemaInterface<TValue> {
  type: string;
  required: [string?];
  properties: TValue;
}

export interface JsonMapperFunction {
  [index: string]: string | number | any;
}

export interface PropertiesInterface<T> {
  [index: string]:
    | ((
        dbProvider: any,
        propertyValue: string | number | [] | PropertiesInterface<T>
      ) => T | JsonMapperFunction | Promise<JsonMapperFunction>)
    | string
    | number;
}
