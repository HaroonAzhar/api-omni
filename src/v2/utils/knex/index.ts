import _Knex from 'knex';

export * from './knex.service';
export * from './knex.module';
export * from './constants';
export type KnexInstance = _Knex;
