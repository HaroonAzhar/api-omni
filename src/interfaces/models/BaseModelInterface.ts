import Knex, { QueryBuilder } from "knex";

import { JsonSchemaInterface } from "./JsonSchemaInterface";

export interface BaseModelInterface {
  jsonSchema(): JsonSchemaInterface<any>;
  tableName(): string;
  insert(): Promise<any>;
  setData(key: string, value: any): void;
  select(fields?: string | Record<string, unknown>): QueryBuilder;
  update(fields?: Record<string, unknown>): QueryBuilder<unknown, number>;
  enableTransaction(trx: Knex.Transaction): BaseModelInterface;
}
