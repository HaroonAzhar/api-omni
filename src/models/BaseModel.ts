import Knex, { QueryBuilder, Raw } from "knex";
import * as Joi from "joi";
import { camelize, underscore } from "inflected";

import { ModelError } from "../errors/ModelError";
import { BaseModelInterface } from "../interfaces/models/BaseModelInterface";
import {
  JsonMapperFunction,
  JsonSchemaInterface,
  PropertiesInterface,
} from "../interfaces/models/JsonSchemaInterface";
import { Dictionary } from "../util/utilityTypes";

export abstract class BaseModel implements BaseModelInterface {
  private static knex: Knex;
  protected _id: number;
  protected _data: PropertiesInterface<Record<string, unknown>> = {};
  protected _jsonObject: PropertiesInterface<Record<string, unknown>> = {};
  protected primaryKey: string | null = null;
  private processedData: [] = [];
  private transaction = false;
  private trx: Knex.Transaction;

  constructor() {
    this.initializeData();
  }

  private processFieldOption(field: string, fieldValue: string) {
    if (fieldValue.indexOf("|") !== -1) {
      const data = fieldValue.split("|");
      data.forEach((dataValue) => {
        if (dataValue === "pk") {
          this.primaryKey = field;
        }
      });
    }
  }

  private initializeData() {
    Object.keys(this.jsonSchema().properties).forEach((value) => {
      this.processFieldOption(
        value,
        this.jsonSchema().properties[value] as string
      );
    });
  }

  isValid() {
    if (!this.tableName() || 0 === this.tableName().length) {
      throw new ModelError("Table name is not set");
    }
  }

  private removeJSONFields() {
    Object.keys(this._jsonObject).forEach(async (value) => {
      if (!this.getJsonMapping().hasOwnProperty(value)) {
        delete this._jsonObject[value];
      }
    });
  }

  private processJSONMapping() {
    return new Promise(async (resolve) => {
      this.removeJSONFields();
      let processed = 0;
      const checkIsProcessed = () => {
        if (processed === Object.keys(this._jsonObject).length) {
          resolve(null);
        }
      };
      checkIsProcessed();
      for (const value of Object.keys(this._jsonObject)) {
        if (this.getJsonMapping().hasOwnProperty(value)) {
          if (typeof this.getJsonMapping()[value] === "function") {
            const data = await (this.getJsonMapping()[value] as (
              dbData: Knex,
              propetyValue: string
            ) => JsonMapperFunction)(
              BaseModel.getKnex(),
              this._jsonObject[value] as string
            );
            if (((data as unknown) as string) === "skip") {
              processed++;
              continue;
            }
            if (data.hasOwnProperty("key"))
              this.setData(data.key as string, data.value);
            processed++;
          } else {
            this.setData(
              this.getJsonMapping()[value] as string,
              this._jsonObject[value]
            );
            processed++;
          }
        }

        checkIsProcessed();
      }
    });
  }

  public async setJsonObject(object: Record<string, any>) {
    this._data = {};
    this._jsonObject = { ...object };
    await this.processJSONMapping();
    return this;
  }

  public data(): PropertiesInterface<Record<string, unknown>> {
    const output: PropertiesInterface<Record<string, unknown>> = {};
    Object.entries(this._data).map(
      ([key, value]) => (output[underscore(key)] = value)
    );
    return output;
  }

  public update(fields?: Record<string, unknown>) {
    const knex = this.transaction ? this.trx : BaseModel.getKnex();
    if (Object.entries(this._data).length === 0) {
      return this.select();
    }
    const updateData = this._data;
    Object.keys(updateData).forEach((val) => {
      //  if (!updateData[val]) delete updateData[val];
    });
    const output: PropertiesInterface<Record<string, unknown>> = {};
    Object.entries(this._data).map(
      ([key, value]) => (output[camelize(key)] = value)
    );
    const query = knex.update(output).into(this.tableName());
    if (this._id) query.where({ [camelize(this.primaryKey)]: this._id });
    return query;
  }

  public delete() {
    const knex = this.transaction ? this.trx : BaseModel.getKnex();
    const query = knex.delete().from(this.tableName());
    if (this._id) query.where({ [camelize(this.primaryKey)]: this._id });
    return query;
  }

  private static validateSchema() {
    Joi.object({
      type: Joi.string().alphanum(),
      required: Joi.array().required(),
      properties: Joi.array().min(1),
    });
  }

  static registerKnex(driver: Knex) {
    Object.defineProperty(this, "knex", {
      writable: true,
      value: driver,
    });
  }

  public makeObjectFromArray<
    T extends Record<string, unknown>,
    S extends string = string
  >(array: Array<keyof T>): Dictionary<T> {
    const obj: Dictionary<T> = {};

    array.forEach((item) => {
      obj[item] = (item as unknown) as S;
    });

    return obj;
  }

  public enableTransaction(trx: Knex.Transaction) {
    this.transaction = true;
    this.trx = trx;
    return this;
  }

  public select(
    fields?: string | Record<string, unknown> | (string | Raw<any>)[]
  ): QueryBuilder {
    const fieldNames = Object.keys(this.jsonSchema().properties).map(
      (propertyName) => `${camelize(propertyName)} as ${propertyName}`
    );
    const knex = BaseModel.getKnex();
    const query = knex
      .select(fields ? fields : fieldNames)
      .from(this.tableName());
    if (this._id) query.where({ [camelize(this.primaryKey)]: this._id });
    return query;
  }

  async insert(): Promise<number[]> {
    const knex = this.transaction ? this.trx : BaseModel.getKnex();
    const output: PropertiesInterface<Record<string, unknown>> = {};
    Object.entries(this._data).map(
      ([key, value]) => (output[camelize(key)] = value)
    );
    const command = knex.insert(output).into(this.tableName());
    return command.returning(camelize(this.primaryKey));
  }

  static getKnex() {
    return this.knex;
  }

  public setData(key: string, value: any) {
    if (!this.jsonSchema().properties.hasOwnProperty(key))
      throw new Error("Cannot find property: " + key);
    this._data[key] = value;
  }

  public clearData() {
    this._data = {};
    return this;
  }

  public get() {
    return true;
  }

  set id(id: number) {
    this._id = id;
  }

  abstract jsonSchema(): JsonSchemaInterface<any>;
  abstract tableName(): string;
  abstract getJsonMapping(): PropertiesInterface<Record<string, unknown>>;
}
