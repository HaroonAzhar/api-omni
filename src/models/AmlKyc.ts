import { BaseModel } from "./BaseModel";
import {
  JsonSchemaInterface,
  PropertiesInterface,
} from "../interfaces/models/JsonSchemaInterface";
import { JsonProps } from "../util/utilityTypes";

type AmlKyc = {
  aml_kyc_id: number;
  status: string;
  fk_applicant_id: number;
  fk_applicant_company_id: number;
  voters_roll: boolean;
  aml: boolean;
  aml_details?: string;
  ccj: boolean;
  ccj_details?: string;
  cifas: boolean;
  cifas_details?: string;
  creditsafe: boolean;
  creditsafe_details?: string;
  google_search: boolean;
  google_search_details?: string;
  experian_risk_score: number;
  proof_of_address: string;
  proof_of_address_details: string;
  additional_proof_of_address: string;
  proof_of_right_to_remain: string;
  proof_of_id: string;
  proof_of_id_details: string;
  proof_of_id_expiry_date: string;
  client_arrival_in_uk: boolean;
  third_party_verification: boolean;
  is_pep: boolean;
  electronic_document_check_satisfactory: boolean;
  is_on_sanctions_list: boolean;
  copies_on_file: boolean;
  adverse_media_check_satisfactory: boolean;
  links_to_high_risk_jurisdiction: string;
  any_suspicion_of_money_laundering: boolean;
  additional_notes: string;
  aml_kyc_company_id: number;
  icaew_membership_confirmed: boolean;
  person_with_interest_verified_as_individual: boolean;
  corporate_structure: string;
  verification_company_is_subsidiary: boolean;
  sedol_no_exchange: string;
  evidence_of_authority_to_act_for_firm: boolean;
  number_of_individuals: string;
  partner_verified_as_individual: boolean;
  evidence_of_trading_address: string;
  purpose_of_partnership: string;
  partnership_agreement_on_file: string;
  person_with_control_verified_as_individual: boolean;
  certificate_of_incorporation: boolean;
  list_of_shareholders: boolean;
  verification_documents_on_file: string;
  shareholders_verified_as_individual: boolean;
  law_society_membership_confirmed: boolean;
  settlor_type: string;
  trustee_type: string;
  controller_type: string;
  source_of_funds_confirmed: boolean;
  creditsafe_clear: boolean;
  creditsafe_clear_details: string;
  individuals_shareholders: boolean;
  company_shareholders: boolean;

  creditsafe_user_comments: string;
  creditsafe_mlro_comments: string;
  creditsafe_mlro_username: string;
  creditsafe_mlro_date: string;
  creditsafe_mlro_state: string;

  icaew_membership_confirmed_user_comments: string;
  icaew_membership_confirmed_mlro_comments: string;
  icaew_membership_confirmed_mlro_username: string;
  icaew_membership_confirmed_mlro_date: string;
  icaew_membership_confirmed_mlro_state: string;

  law_society_membership_confirmed_user_comments: string;
  law_society_membership_confirmed_mlro_comments: string;
  law_society_membership_confirmed_mlro_username: string;
  law_society_membership_confirmed_mlro_date: string;
  law_society_membership_confirmed_mlro_state: string;

  certificate_of_incorporation_user_comments: string;
  certificate_of_incorporation_mlro_comments: string;
  certificate_of_incorporation_mlro_username: string;
  certificate_of_incorporation_mlro_date: string;
  certificate_of_incorporation_mlro_state: string;

  list_of_shareholders_user_comments: string;
  list_of_shareholders_mlro_comments: string;
  list_of_shareholders_mlro_username: string;
  list_of_shareholders_mlro_date: string;
  list_of_shareholders_mlro_state: string;

  shareholders_verified_as_individual_user_comments: string;
  shareholders_verified_as_individual_mlro_comments: string;
  shareholders_verified_as_individual_mlro_username: string;
  shareholders_verified_as_individual_mlro_date: string;
  shareholders_verified_as_individual_mlro_state: string;

  person_with_control_verified_as_individual_user_comments: string;
  person_with_control_verified_as_individual_mlro_comments: string;
  person_with_control_verified_as_individual_mlro_username: string;
  person_with_control_verified_as_individual_mlro_date: string;
  person_with_control_verified_as_individual_mlro_state: string;

  person_with_interest_verified_as_individual_user_comments: string;
  person_with_interest_verified_as_individual_mlro_comments: string;
  person_with_interest_verified_as_individual_mlro_username: string;
  person_with_interest_verified_as_individual_mlro_date: string;
  person_with_interest_verified_as_individual_mlro_state: string;

  partner_verified_as_individual_user_comments: string;
  partner_verified_as_individual_mlro_comments: string;
  partner_verified_as_individual_mlro_username: string;
  partner_verified_as_individual_mlro_date: string;
  partner_verified_as_individual_mlro_state: string;

  partnership_agreement_on_file_user_comments: string;
  partnership_agreement_on_file_mlro_comments: string;
  partnership_agreement_on_file_mlro_username: string;
  partnership_agreement_on_file_mlro_date: string;
  partnership_agreement_on_file_mlro_state: string;

  company_shareholders_user_comments: string;
  company_shareholders_mlro_comments: string;
  company_shareholders_mlro_username: string;
  company_shareholders_mlro_date: string;
  company_shareholders_mlro_state: string;

  individuals_shareholders_user_comments: string;
  individuals_shareholders_mlro_comments: string;
  individuals_shareholders_mlro_username: string;
  individuals_shareholders_mlro_date: string;
  individuals_shareholders_mlro_state: string;

  adverse_media_check_satisfactory_user_comments: string;
  adverse_media_check_satisfactory_mlro_comments: string;
  adverse_media_check_satisfactory_mlro_username: string;
  adverse_media_check_satisfactory_mlro_date: string;
  adverse_media_check_satisfactory_mlro_state: string;

  copies_on_file_user_comments: string;
  copies_on_file_mlro_comments: string;
  copies_on_file_mlro_username: string;
  copies_on_file_mlro_date: string;
  copies_on_file_mlro_state: string;

  electronic_document_check_satisfactory_user_comments: string;
  electronic_document_check_satisfactory_mlro_comments: string;
  electronic_document_check_satisfactory_mlro_username: string;
  electronic_document_check_satisfactory_mlro_date: string;
  electronic_document_check_satisfactory_mlro_state: string;

  is_on_sanctions_list_user_comments: string;
  is_on_sanctions_list_mlro_comments: string;
  is_on_sanctions_list_mlro_username: string;
  is_on_sanctions_list_mlro_date: string;
  is_on_sanctions_list_mlro_state: string;

  is_pep_user_comments: string;
  is_pep_mlro_comments: string;
  is_pep_mlro_username: string;
  is_pep_mlro_date: string;
  is_pep_mlro_state: string;

  creditsafe_clear_user_comments: string;
  creditsafe_clear_mlro_comments: string;
  creditsafe_clear_mlro_username: string;
  creditsafe_clear_mlro_date: string;
  creditsafe_clear_mlro_state: string;

  links_to_high_risk_jurisdiction_user_comments: string;
  links_to_high_risk_jurisdiction_mlro_comments: string;
  links_to_high_risk_jurisdiction_mlro_username: string;
  links_to_high_risk_jurisdiction_mlro_date: string;
  links_to_high_risk_jurisdiction_mlro_state: string;

  any_suspicion_of_money_laundering_user_comments: string;
  any_suspicion_of_money_laundering_mlro_comments: string;
  any_suspicion_of_money_laundering_mlro_username: string;
  any_suspicion_of_money_laundering_mlro_date: string;
  any_suspicion_of_money_laundering_mlro_state: string;

  proof_of_right_to_remain_user_comments: string;
  proof_of_right_to_remain_mlro_comments: string;
  proof_of_right_to_remain_mlro_username: string;
  proof_of_right_to_remain_mlro_date: string;
  proof_of_right_to_remain_mlro_state: string;
};

const properties: Array<keyof AmlKyc> = [
  "fk_applicant_id",
  "status",
  "voters_roll",
  "aml",
  "aml_details",
  "ccj",
  "ccj_details",
  "cifas",
  "cifas_details",
  "creditsafe",
  "creditsafe_details",
  "google_search",
  "google_search_details",
  "experian_risk_score",
  "proof_of_address",
  "proof_of_address_details",
  "additional_proof_of_address",
  "proof_of_right_to_remain",
  "proof_of_id",
  "proof_of_id_details",
  "proof_of_id_expiry_date",
  "client_arrival_in_uk",
  "third_party_verification",
  "is_pep",
  "electronic_document_check_satisfactory",
  "is_on_sanctions_list",
  "copies_on_file",
  "adverse_media_check_satisfactory",
  "links_to_high_risk_jurisdiction",
  "any_suspicion_of_money_laundering",
  "additional_notes",
  "corporate_structure",
  "fk_applicant_id",
  "verification_company_is_subsidiary",
  "sedol_no_exchange",
  "evidence_of_authority_to_act_for_firm",
  "number_of_individuals",
  "partner_verified_as_individual",
  "icaew_membership_confirmed",
  "person_with_interest_verified_as_individual",
  "evidence_of_trading_address",
  "purpose_of_partnership",
  "partnership_agreement_on_file",
  "person_with_control_verified_as_individual",
  "certificate_of_incorporation",
  "list_of_shareholders",
  "verification_documents_on_file",
  "shareholders_verified_as_individual",
  "law_society_membership_confirmed",
  "settlor_type",
  "trustee_type",
  "controller_type",
  "source_of_funds_confirmed",
  "fk_applicant_company_id",
  "creditsafe_clear",
  "creditsafe_clear_details",
  "individuals_shareholders",
  "company_shareholders",

  "creditsafe_user_comments",
  "creditsafe_mlro_comments",
  "creditsafe_mlro_username",
  "creditsafe_mlro_date",
  "creditsafe_mlro_state",

  "icaew_membership_confirmed_user_comments",
  "icaew_membership_confirmed_mlro_comments",
  "icaew_membership_confirmed_mlro_username",
  "icaew_membership_confirmed_mlro_date",
  "icaew_membership_confirmed_mlro_state",

  "law_society_membership_confirmed_user_comments",
  "law_society_membership_confirmed_mlro_comments",
  "law_society_membership_confirmed_mlro_username",
  "law_society_membership_confirmed_mlro_date",
  "law_society_membership_confirmed_mlro_state",

  "certificate_of_incorporation_user_comments",
  "certificate_of_incorporation_mlro_comments",
  "certificate_of_incorporation_mlro_username",
  "certificate_of_incorporation_mlro_date",
  "certificate_of_incorporation_mlro_state",

  "list_of_shareholders_user_comments",
  "list_of_shareholders_mlro_comments",
  "list_of_shareholders_mlro_username",
  "list_of_shareholders_mlro_date",
  "list_of_shareholders_mlro_state",

  "shareholders_verified_as_individual_user_comments",
  "shareholders_verified_as_individual_mlro_comments",
  "shareholders_verified_as_individual_mlro_username",
  "shareholders_verified_as_individual_mlro_date",
  "shareholders_verified_as_individual_mlro_state",

  "person_with_control_verified_as_individual_user_comments",
  "person_with_control_verified_as_individual_mlro_comments",
  "person_with_control_verified_as_individual_mlro_username",
  "person_with_control_verified_as_individual_mlro_date",
  "person_with_control_verified_as_individual_mlro_state",

  "person_with_interest_verified_as_individual_user_comments",
  "person_with_interest_verified_as_individual_mlro_comments",
  "person_with_interest_verified_as_individual_mlro_username",
  "person_with_interest_verified_as_individual_mlro_date",
  "person_with_interest_verified_as_individual_mlro_state",

  "partner_verified_as_individual_user_comments",
  "partner_verified_as_individual_mlro_comments",
  "partner_verified_as_individual_mlro_username",
  "partner_verified_as_individual_mlro_date",
  "partner_verified_as_individual_mlro_state",

  "partnership_agreement_on_file_user_comments",
  "partnership_agreement_on_file_mlro_comments",
  "partnership_agreement_on_file_mlro_username",
  "partnership_agreement_on_file_mlro_date",
  "partnership_agreement_on_file_mlro_state",

  "company_shareholders_user_comments",
  "company_shareholders_mlro_comments",
  "company_shareholders_mlro_username",
  "company_shareholders_mlro_date",
  "company_shareholders_mlro_state",

  "individuals_shareholders_user_comments",
  "individuals_shareholders_mlro_comments",
  "individuals_shareholders_mlro_username",
  "individuals_shareholders_mlro_date",
  "individuals_shareholders_mlro_state",

  "adverse_media_check_satisfactory_user_comments",
  "adverse_media_check_satisfactory_mlro_comments",
  "adverse_media_check_satisfactory_mlro_username",
  "adverse_media_check_satisfactory_mlro_date",
  "adverse_media_check_satisfactory_mlro_state",

  "copies_on_file_user_comments",
  "copies_on_file_mlro_comments",
  "copies_on_file_mlro_username",
  "copies_on_file_mlro_date",
  "copies_on_file_mlro_state",

  "electronic_document_check_satisfactory_user_comments",
  "electronic_document_check_satisfactory_mlro_comments",
  "electronic_document_check_satisfactory_mlro_username",
  "electronic_document_check_satisfactory_mlro_date",
  "electronic_document_check_satisfactory_mlro_state",

  "is_on_sanctions_list_user_comments",
  "is_on_sanctions_list_mlro_comments",
  "is_on_sanctions_list_mlro_username",
  "is_on_sanctions_list_mlro_date",
  "is_on_sanctions_list_mlro_state",

  "is_pep_user_comments",
  "is_pep_mlro_comments",
  "is_pep_mlro_username",
  "is_pep_mlro_date",
  "is_pep_mlro_state",

  "creditsafe_clear_user_comments",
  "creditsafe_clear_mlro_comments",
  "creditsafe_clear_mlro_username",
  "creditsafe_clear_mlro_date",
  "creditsafe_clear_mlro_state",

  "links_to_high_risk_jurisdiction_user_comments",
  "links_to_high_risk_jurisdiction_mlro_comments",
  "links_to_high_risk_jurisdiction_mlro_username",
  "links_to_high_risk_jurisdiction_mlro_date",
  "links_to_high_risk_jurisdiction_mlro_state",

  "any_suspicion_of_money_laundering_user_comments",
  "any_suspicion_of_money_laundering_mlro_comments",
  "any_suspicion_of_money_laundering_mlro_username",
  "any_suspicion_of_money_laundering_mlro_date",
  "any_suspicion_of_money_laundering_mlro_state",

  "proof_of_right_to_remain_user_comments",
  "proof_of_right_to_remain_mlro_comments",
  "proof_of_right_to_remain_mlro_username",
  "proof_of_right_to_remain_mlro_date",
  "proof_of_right_to_remain_mlro_state",
];

export default class AmlKycModel extends BaseModel {
  tableName(): string {
    return "Origination.AmlKyc";
  }

  getJsonMapping(): PropertiesInterface<string> {
    return this.makeObjectFromArray<AmlKyc>(properties);
  }

  jsonSchema(): JsonSchemaInterface<JsonProps<AmlKyc>> {
    return {
      type: "object",
      required: [],
      properties: {
        ...this.makeObjectFromArray<AmlKyc>(properties),
        aml_kyc_id: "id|pk",
      },
    };
  }
}
