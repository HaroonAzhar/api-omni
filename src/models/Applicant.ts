import Knex from "knex";

import { BaseModel } from "./BaseModel";
import {
  JsonSchemaInterface,
  PropertiesInterface,
} from "../interfaces/models/JsonSchemaInterface";
import Security from "./Dip/Security";
import ContactValue from "./Dip/ContactValue";
import ApplicationStepStatusType from "./Application/ApplicationStepStatusType";

export enum ApplicantType {
  INDIVIDUAL = "individual",
  COMPANY = "company",
}

export default class Applicant extends BaseModel {
  getJsonMapping(): PropertiesInterface<Record<string, unknown>> {
    return {
      title: "title",
      forename: "forename",
      middle_name: "middle_name",
      surname: "surname",
      other_name: "other_name",
      date_of_birth: "date_of_birth",
      city_of_birth: "city_of_birth",
      country_of_birth: "country_of_birth",
      insurance_number: "insurance_number",
      nationality: "nationality",
      second_nationality: "second_nationality",
      has_dual_nationality: "has_dual_nationality",
      permanent_resident: "permanent_resident",
      marital_status: "marital_status",
      marital_other_value: "marital_other_value",
      declaration: "declaration",
      signature: "signature",
      date_of_declaration: "date_of_declaration",
      date_of_signature: "date_of_signature",
      mothers_maiden_name: "mothers_maiden_name",
      uk_residential_status: "uk_residential_status",
      information_regarding_property_residence:
        "information_regarding_property_residence",
      personal_data: async (db: Knex, value: PropertiesInterface<string>) => {
        for (const propKey in value) {
          if (value.hasOwnProperty(propKey))
            this.setData(propKey, value[propKey]);
        }
        return "skip";
      },
      contact: async (db: Knex, value: any) => {
        const model = new ContactValue();

        for (const propKey in value) {
          if (value.hasOwnProperty(propKey) && propKey !== "contact_id")
            model.setData(propKey, value[propKey]);
        }

        if (value.hasOwnProperty("contact_id")) {
          model.id = value.contact_id;
          await model.update();
          return true;
        }

        const lastInsertId = await model.insert();
        return {
          key: "fk_contact_id",
          value: lastInsertId[0],
        };
      },

      addresses: async (db: Knex, value: []) => {
        const securityModel = new Security();
        const deleteCondition = {
          FkApplicantId: this._id,
          AddressType: "individual",
        };
        const data = await db
          .select()
          .from("Origination.ApplicantAddressMapping")
          .where(deleteCondition);
        if (data.length) {
          await db
            .delete()
            .from("Origination.ApplicantAddressMapping")
            .where(deleteCondition);
          for (const dataValue of data) {
            await securityModel.delete().where({
              SecurityId: dataValue.FkAddressId,
            });
          }
        }

        for (const address of value) {
          securityModel.setData(
            "security_address_line_1",
            address["address_line_1"]
          );
          securityModel.setData(
            "security_address_line_2",
            address["address_line_2"]
          );
          securityModel.setData("security_town_city", address["city"]);
          securityModel.setData("security_postcode", address["postcode"]);
          securityModel.setData("security_country", address["country"]);
          securityModel.setData(
            "how_long_here_months",
            address["how_long_here_months"]
          );
          securityModel.setData(
            "how_long_here_years",
            address["how_long_here_years"]
          );

          const insertedId = await securityModel.insert();

          await db.into("Origination.ApplicantAddressMapping").insert({
            AddressType: "individual",
            FkApplicantId: this._id,
            FkAddressId: insertedId,
          });
        }

        return {};
      },
      status: async (db: Knex, value: string) => {
        const status = new ApplicationStepStatusType();
        const data = await status
          .select()
          .where({ ApplicationStepStatusType: value });
        if (!data.length) throw new Error("Cannot find status:" + value);
        return {
          key: "fk_status_id",
          value: data[0].application_step_status_id,
        };
      },
      date_edited: "date_edited",
      links: "links",
      is_deleted: "is_deleted",
      fk_shared_contact_id: "fk_shared_contact_id",
    };
  }

  jsonSchema(): JsonSchemaInterface<any> {
    return {
      type: "object",
      required: [],
      properties: {
        applicant_id: "number|pk",
        title: "string",
        forename: "string",
        middle_name: "string",
        surname: "string",
        other_name: "string",
        date_of_birth: "string",
        city_of_birth: "string",
        country_of_birth: "string",
        insurance_number: "string",
        nationality: "string",
        second_nationality: "string",
        has_dual_nationality: "string",
        permanent_resident: "string",
        marital_status: "string",
        marital_other_value: "string",
        declaration: "string",
        signature: "string",
        date_of_declaration: "string",
        date_of_signature: "string",
        mothers_maiden_name: "string",
        uk_residential_status: "string",
        information_regarding_property_residence: "string",
        fk_case_id: "number",
        fk_contact_id: "number",
        fk_applicant_credit_history_id: "number",
        fk_status_id: "number",
        date_edited: "string",
        links: "string",
        is_deleted: "boolean",
        fk_shared_contact_id: "number",
      },
    };
  }

  tableName(): string {
    return "Origination.Applicant";
  }
}
