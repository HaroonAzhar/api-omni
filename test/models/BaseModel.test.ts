import * as Knex from "knex";

import {
  JsonSchemaInterface,
  PropertiesInterface,
} from "../../src/interfaces/models/JsonSchemaInterface";
import { BaseModel } from "../../src/models/BaseModel";

class SampleModel extends BaseModel {
  tableName(): string {
    return "";
  }

  jsonSchema(): JsonSchemaInterface<any> {
    return {
      type: "object",
      required: [],
      properties: {
        sample_pk: "number|pk",
        sample: "string",
        sample_fake: "string",
        sample_fake3: "string",
      },
    };
  }

  getJsonMapping(): PropertiesInterface<any> {
    return {
      sampleNonRelatedField: "sample",
      nextField: "sample_fake",
      otherFakeField: (db: Knex, propertyValue) => {
        db.select().from("faketable");
        return {
          key: "sample_fake3",
          value: propertyValue,
        };
      },
    };
  }
}

describe("BaseModel TestCase", () => {
  it("should initialize knex instance", () => {
    BaseModel.registerKnex({
      sample: true,
    } as any);

    expect(BaseModel.getKnex()).toMatchObject({
      sample: true,
    });
  });

  it("should throw error if table not set", () => {
    expect(() => {
      const model = new SampleModel();
      model.isValid();
    }).toThrowErrorMatchingSnapshot();
  });

  it("should find primary key", () => {
    const model = new SampleModel();
    const primaryKey = (model as any).primaryKey;
    expect(primaryKey).toBe("sample_pk");
  });

  it("primary key should not be visible for insert data", () => {
    const insertMock: any = jest.fn(() => {
      return {
        into: () => {
          return {
            returning: () => {
              return 1;
            },
          };
        },
      };
    });
    BaseModel.registerKnex({
      sample: true,
      insert: insertMock,
    } as any);
    const model = new SampleModel();
    model.tableName = () => "sample";
    model.setData("sample", "test");
    model.insert();

    expect(insertMock.mock.calls[0][0]).toEqual({
      Sample: "test",
    });
  });

  it("should map fields from json and update", () => {
    const updateMock = jest.fn(() => {
      {
        return {
          into: () => {
            return {
              returning: () => {
                return 1;
              },
            };
          },
        };
      }
    });

    BaseModel.registerKnex({
      update: updateMock,
    } as any);

    const model = new SampleModel();
    const json = {
      sampleNonRelatedField: "somefakevalue",
      nextField: "nextfake",
    };

    model.setJsonObject(json);
    model.update();
    expect(model.data().sample).toBe("somefakevalue");
    expect(model.data().sample_fake).toBe("nextfake");
  });

  it("should update data", () => {
    const updateMock = jest.fn(() => {
      {
        return {
          into: () => {
            return {
              returning: () => {
                return 1;
              },
            };
          },
        };
      }
    });

    BaseModel.registerKnex({
      update: updateMock,
    } as any);

    const model = new SampleModel();
    model.setData("sample", "test");
    model.update();
    const calls: any = updateMock.mock.calls;
    expect(calls[0][0]).toMatchObject({
      Sample: "test",
    });
  });

  it("should update if null value", () => {
    const updateMock = jest.fn(() => {
      {
        return {
          into: () => {
            return {
              returning: () => {
                return 1;
              },
            };
          },
        };
      }
    });

    BaseModel.registerKnex({
      update: updateMock,
    } as any);

    const model = new SampleModel();
    model.setData("sample", "test");
    model.setData("sample", null);
    model.update();
    const calls: any = updateMock.mock.calls;
    expect(calls[0][0]).toMatchObject({
      Sample: null,
    });
  });

  it("should mapper recognize function with db", async () => {
    const updateMock = jest.fn(() => {
      {
        return {
          into: () => {
            return {
              returning: () => {
                return 1;
              },
            };
          },
        };
      }
    });

    const model = new SampleModel();
    const fromMock = jest.fn();
    const selectMock = jest.fn(() => {
      return {
        from: fromMock,
      };
    });
    BaseModel.registerKnex({
      select: selectMock,
      update: updateMock,
    } as any);

    await model.setJsonObject({
      otherFakeField: "samplefake222",
    });

    model.update();

    expect(selectMock.mock.calls.length).toBe(1);
    expect(fromMock.mock.calls[0][0]).toBe("faketable");
    expect(model.data().sample_fake3).toBe("samplefake222");
  });

  /*    it("should insert data", () => {
        const insertMock = jest.fn();
        BaseModel.registerKnex({
            sample: true,
            table: () => {
                return {
                    insert: insertMock
                };
            }
        } as any);
        const model = new SampleModel();
        model.tableName = () => "sample";
        model.setData("sample", "test");
        model.insert();

        expect(insertMock.mock.calls[0][0]).toMatchObject({
            sample: "test",
            sampleFake: null
        });
    });*/
});
