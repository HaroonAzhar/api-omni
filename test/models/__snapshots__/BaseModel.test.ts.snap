// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseModel TestCase should throw error if table not set 1`] = `"Table name is not set"`;
